### PR TITLE
Instrument State, Layer, RenderObject and Element.

### PR DESCRIPTION
**Pull request by polina-c**
See https://github.com/flutter/devtools/issues/3951

### CODE REVIEW

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 1]
```diff
@@ -463,10 +467,10 @@ class _MergingListenable extends Listenable {
 class ValueNotifier<T> extends ChangeNotifier implements ValueListenable<T> {
   /// Creates a [ChangeNotifier] that wraps this value.
   ValueNotifier(this._value) {
-    if (kFlutterMemoryAllocationsEnabled) {
+    if (kFlutterMemoryAllocationsEnabled && MemoryAllocations.instance.hasListeners) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/change_notifier.dart:470]: Why do these calls need to be guarded by `MemoryAllocations.instance.hasListeners`? It makes this a little more cumbersome to use and it is another thing that's easy to forget when using this API. If this guard is indeed necessary, please update the documentation on dispatchObjectEvent and maybe assert in that method that it is only called when there are listeners?

File: packages/flutter/test/foundation/memory_allocations_test.dart [hunk 2]
```diff
@@ -150,7 +152,7 @@ void main() {
 
   test('publishers in Flutter dispatch events in debug mode', () async {
     int eventCount = 0;
-    void listener(ObjectEvent event) => eventCount++;
+    void listener(ObjectEvent event) =>  eventCount++;
```
Comment (goderbauer) [inline packages/flutter/test/foundation/memory_allocations_test.dart:155]: nit: remove the extra space?

File: packages/flutter/test/foundation/memory_allocations_test.dart [hunk 3]
```diff
@@ -160,20 +162,97 @@ void main() {
     _checkSdkHandlersNotSet();
     expect(ma.hasListeners, isFalse);
   });
+
+  testWidgets('State dispatch events in debug mode', (WidgetTester tester) async {
```
Comment (goderbauer) [inline packages/flutter/test/foundation/memory_allocations_test.dart:166]: This test probably belongs under `/test/widgets` and not foundation.

File: packages/flutter/test/foundation/memory_allocations_test.dart [hunk 4]
```diff
@@ -160,20 +162,97 @@ void main() {
...
+        stateCreated = true;
+      }
+      if (event is ObjectDisposed && event.object is State) {
+        stateDisposed = true;
+      }
+    }
+    ma.addListener(listener);
+
+    await tester.pumpWidget(const _TestStatefulWidget());
+    await tester.pumpWidget(const SizedBox.shrink());
```
Comment (goderbauer) [inline packages/flutter/test/foundation/memory_allocations_test.dart:181]: maybe assert in between the pumps that the state was created but not disposed yet?

File: packages/flutter/test/foundation/memory_allocations_test.dart (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/test/foundation/memory_allocations_test.dart:166]: Same below, the element test also belongs to `test/widgets`. The Layer and RO test belongs to `test/rendering`.

File: packages/flutter/test_release/foundation/memory_allocations_test.dart [hunk 5]
```diff
@@ -40,34 +43,104 @@ void main() {
...
+  void performLayout() {}
+
+  @override
+  void performResize() {}
+
+  @override
+  Rect get semanticBounds => throw UnimplementedError();
+}
+
+class _MyStateFulWidget extends StatefulWidget {
```
Comment (goderbauer) [inline packages/flutter/test_release/foundation/memory_allocations_test.dart:86]: Spelling here and elsewhere: StateFul -> Stateful

File: packages/flutter/test_release/foundation/memory_allocations_test.dart [hunk 6]
```diff
@@ -2,9 +2,11 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-import 'dart:ui';
+import 'dart:ui' as ui;
```
Comment (goderbauer) [inline packages/flutter/test_release/foundation/memory_allocations_test.dart:5]: The same comment about test organization also applies here.

File: packages/flutter/lib/src/foundation/change_notifier.dart (hunk 1 above)
Comment (polina-c) [reply packages/flutter/lib/src/foundation/change_notifier.dart:470]: It is not necessary, but I want to save on object creation.
I ended up moving object creation to MemoryAllocations

File: packages/flutter/test/foundation/memory_allocations_test.dart (hunk 3 above)
Comment (polina-c) [reply packages/flutter/test/foundation/memory_allocations_test.dart]: done

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 7]
```diff
@@ -464,11 +464,11 @@ class ValueNotifier<T> extends ChangeNotifier implements ValueListenable<T> {
   /// Creates a [ChangeNotifier] that wraps this value.
   ValueNotifier(this._value) {
     if (kFlutterMemoryAllocationsEnabled) {
-      MemoryAllocations.instance.dispatchObjectEvent(ObjectCreated(
+      MemoryAllocations.instance.dispatchObjectCreated(
         library: _flutterFoundationLibrary,
-        className: 'ValueNotifier',
+        className: '$ValueNotifier',
```
Comment (dnfield) [inline packages/flutter/lib/src/foundation/change_notifier.dart:469]: Do these end up being equivalent to `runtimeType.toString` under the hood?

Comment (dnfield) [reply packages/flutter/lib/src/foundation/change_notifier.dart:469]: Looks like, no, it's emitted as a `TypeLiteralConstant` which should be fine...

**Review by dnfield — APPROVED** [review]

**Pull request merged**